src/rgw/driver/sfs/sqlite/dbconn.h

```diff
@@ -202,24 +204,64 @@ inline auto _make_storage(const std::string& path) {
 
 using Storage = decltype(_make_storage(""));
 
+static void sqlite_error_callback(void* ctx, int error_code, const char* msg) {
+  const auto cct = static_cast<CephContext*>(ctx);
+  lderr(cct) << "[SQLITE] (" << error_code << ") " << msg << dendl;
+}
+
+static int sqlite_profile_callback(
+    unsigned int reason, void* ctx, void* vstatement, void* runtime_ns
+) {
+  const auto cct = static_cast<CephContext*>(ctx);
+
+  if (reason == SQLITE_TRACE_PROFILE) {
+    const int64_t runtime = *static_cast<int64_t*>(runtime_ns) / 1000 / 1000;
+    sqlite3_stmt* statement = static_cast<sqlite3_stmt*>(vstatement);
+    const std::unique_ptr<char, void (*)(char*)> sql(
+        sqlite3_expanded_sql(statement),
+        [](char* p) { sqlite3_free(static_cast<void*>(p)); }
```

Comment:
is it possible `p` to be null?

Reply:
ok, the sql docs also say that:  "The sqlite3_free() routine is a no-op if is called with a NULL pointer. Passing a NULL pointer to sqlite3_free() is harmless", so double safe.
By the way, perhaps could be worth adding a `if(sql)` before unwrapping `sql`  in the ` lderr` statement ?

Reply:
Last push adds an `if(sql)` with fallback to `sqlite3_sql`, which returns a const char* to something prepared during statement execution.